tokio/tests/io_poll_aio.rs

```diff
@@ -0,0 +1,147 @@
+#![warn(rust_2018_idioms)]
+#![cfg(all(target_os = "freebsd", feature = "aio"))]
+
+use futures_test::task::noop_context;
+use mio_aio::{AioCb, AioFsyncMode, LioCb};
+use std::{
+    io::{Read, Seek, SeekFrom},
+    os::unix::io::{AsRawFd, RawFd},
+    task::Poll,
+    time::{Duration, Instant},
+};
+use tempfile::tempdir;
+use tokio::io::{AioSource, PollAio};
+use tokio_test::{assert_pending, assert_ready_ok};
+
+fn rt() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_io()
+        .build()
+        .unwrap()
+}
+
+struct WrappedAioCb<'a>(AioCb<'a>);
+impl<'a> AioSource for WrappedAioCb<'a> {
+    fn register(&mut self, kq: RawFd, token: usize) {
+        self.0.register_raw(kq, token)
+    }
+    fn deregister(&mut self) {
+        self.0.deregister_raw()
+    }
+}
+
+struct WrappedLioCb<'a>(LioCb<'a>);
+impl<'a> AioSource for WrappedLioCb<'a> {
+    fn register(&mut self, kq: RawFd, token: usize) {
+        self.0.register_raw(kq, token)
+    }
+    fn deregister(&mut self) {
+        self.0.deregister_raw()
+    }
+}
+
+mod aio {
+    use super::*;
+
+    /// aio_fsync is the simplest AIO operation.  This test ensures that Tokio
+    /// can register an aiocb, and set its readiness.
```

Comment:
For both tests, I'd like to see something that performs more than one operation to verify that the readiness flag behaves properly.

Reply:
For non-lio-listio operations there really isn't any such thing as performing more than one operation.  The kernel forgets about the aiocb as soon as you call `aio_return` to get its result.  You could change some fields and resubmit it, but that's equivalent to creating a new `aiocb`.

For lio_listio, it's hard to write a small test that requires `clear_readiness`.  The only way that would ever be required would be due to a race or a system resource limitation.  Here's a test that deliberately hits a system resource limitation
https://github.com/asomers/tokio-file/blob/master/tests/lio_listio_incomplete.rs

Reply:
If you already have a test for it, it would be nice to include it in Tokio as well.

Reply:
Ok, but it will take a while to adapt, to make it run without all of the other stuff in tokio-file.

Reply:
I think it's important to have at least one test that includes a call to `clear_readiness`.